src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,464 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "textUtils.h"
+#include "uiScreens.h"
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+#define TRACE_WITH_CTX(message, ...) TRACE(message "level = %u, remaining scripts = %u", ##__VA_ARGS__, ctx->level, ctx->complexScripts[ctx->level].remainingScripts)
+#define SAFE_SPRINTF(str, format, ...) {\
+		explicit_bzero(str, SIZEOF(str));\
+		snprintf(str, SIZEOF(str), format, __VA_ARGS__);\
+		ASSERT(strlen(str) + 1 < SIZEOF(str));\
+	}
+
+static inline bool areMoreScriptsExpected()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->complexScripts[ctx->level].remainingScripts > 0;
+}
+
+static inline bool isComplexScriptFinished()
+{
+	return ctx->level > 0 && ctx->complexScripts[ctx->level].remainingScripts == 0;
+}
+
+static inline void complexScriptFinished()
+{
+	while (isComplexScriptFinished()) {
+		ASSERT(ctx->level > 0);
+		ctx->level--;
+
+		ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+		ctx->complexScripts[ctx->level].remainingScripts--;
+
+		TRACE_WITH_CTX("complex script finished, ");
+	}
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+
+	if (isComplexScriptFinished()) {
+		complexScriptFinished();
+	}
+}
+
+// UI
+
+const char ui_native_script_header[7][27] = {"Script - key path", "Script - key", "Script - ALL", "Script - ANY", "Script - N of K", "Script - invalid before", "Script - invalid hereafter"};
+
+#define ASSERT_UI_SCRIPT_TYPE_SANITY() ASSERT(ctx->ui_scriptType >= UI_SCRIPT_PUBKEY_PATH && ctx->ui_scriptType <= UI_SCRIPT_INVALID_HEREAFTER)
+#define HEADER ui_native_script_header[ctx->ui_scriptType]
+
+static uint8_t _getScriptLevelForPosition()
+{
+	// For complex scripts we reduce the current level by 1
+	// Because they already have the level increased by 1
+	int levelOffset = ctx->ui_scriptType == UI_SCRIPT_ALL
+	                  || ctx->ui_scriptType == UI_SCRIPT_ANY
+	                  || ctx->ui_scriptType == UI_SCRIPT_N_OF_K
+	                  ? -1 : 0;
+	ASSERT(ctx->level + levelOffset >= 0);
```

Comment:
I'm a bit uncomfortable with relying on the implicit conversion rules here. uint8_t + int will be int as far as I know, but it seems a bit risky that it might become an unsigned type by some accident or just not paying attention. And then the comparison with 0 is worthless.

Reply:
I tried to come up with a better solution:
```c
	uint8_t levelOffset = ctx->ui_scriptType == UI_SCRIPT_ALL
	                   || ctx->ui_scriptType == UI_SCRIPT_ANY
	                   || ctx->ui_scriptType == UI_SCRIPT_N_OF_K
	                   ? 1 : 0;
	ASSERT(levelOffset == 0 || ctx->level > 0);
	return ctx->level - levelOffset;
```
I switched the type to unsigned int, and the assertion requires the offset to be either 0, or the level to be bigger than 0, which for me seems reasonable. What do you think?

Reply:
I'm OK with this.